src/main/scala/io/iohk/ethereum/ledger/Ledger.scala

```diff
@@ -0,0 +1,249 @@
+package io.iohk.ethereum.ledger
+
+import akka.util.ByteString
+import io.iohk.ethereum.db.storage.NodeStorage
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.utils.Logger
+import io.iohk.ethereum.vm.{GasFee, _}
+
+object Ledger extends Logger {
+
+  type PR = ProgramResult[InMemoryWorldStateProxy, InMemoryWorldStateProxyStorage]
+  type ExecResult = (InMemoryWorldStateProxy, BigInt)
+
+  def executeBlock(
+    block: Block,
+    storages: BlockchainStorages,
+    stateStorage: NodeStorage): Unit = {
+
+    val blockError = validateBlockBeforeExecution(block)
+    if (blockError.isEmpty) {
+      log.debug(s"About to execute txs from block ${block.header}")
+      val (resultingWorldStateProxy, gasUsed) = executeBlockTransactions(block, storages, stateStorage)
+      log.debug(s"All txs from block ${block.header} were executed")
+
+      val worldToPersist = payBlockReward(block, resultingWorldStateProxy)
+      val afterExecutionBlockError = validateBlockAfterExecution(block, worldToPersist)
+      if (afterExecutionBlockError.isEmpty) {
+        InMemoryWorldStateProxy.persistIfHashMatches(block.header.stateRoot, worldToPersist)
+        log.debug(s"Block ${block.header} txs state changes persisted")
+      } else throw new RuntimeException(afterExecutionBlockError.get)
+
+    } else throw new RuntimeException(blockError.get)
+  }
+
+  /**
+    * This function runs transaction
+    *
+    * @param block
+    * @param storages
+    * @param stateStorage
+    */
+  private def executeBlockTransactions(
+    block: Block,
+    storages: BlockchainStorages,
+    stateStorage: NodeStorage):
+  ExecResult = {
+    val blockchain = BlockchainImpl(storages)
+    val initialWorldStateProxy = InMemoryWorldStateProxy(storages, stateStorage,
+      blockchain.getBlockHeaderByHash(block.header.hash).map(_.stateRoot)
+    )
+    block.body.transactionList.foldLeft[ExecResult](initialWorldStateProxy -> 0) {
+      case ((worldStateProxy, acumGas), stx) =>
+        val result: Either[String, PR] = for {
+          _ <- validateTransaction(stx, worldStateProxy, 0, block)
+          worldStateProxy1 = updateAccountBeforeExecution(stx, worldStateProxy)
+          result <- execute(stx, block.header, worldStateProxy1)
+        } yield result
+
+        result match {
+          case Right(theResult) =>
+            val gasUsed = stx.tx.gasLimit - theResult.gasRemaining
+            val refundGasFn = {
+              if (theResult.gasRemaining > 0) (refundGas _).curried(Address(block.header.beneficiary))(gasUsed)
+              else identity[InMemoryWorldStateProxy] _
+            }
+            val payBeneficiariesFn = (payForGasUsedToBeneficiary _).curried(stx)
+            val deleteAccountsFn = (deleteAccounts _).curried(theResult.addressesToDelete)
+            (refundGasFn andThen payBeneficiariesFn andThen deleteAccountsFn) (theResult.world) -> (gasUsed + acumGas)
```

Comment:
also 
```
(refundGasFn andThen payBeneficiariesFn andThen deleteAccountsFn) (theResult.world) -> (gasUsed + acumGas)
```
looks nice but is confusing
maybe (endWorldState is just example name)
```
val endWorldState = (refundGasFn andThen payBeneficiariesFn andThen deleteAccountsFn) (theResult.world)
endWorldState -> (gasUsed + acumGas)
```